doc: add parasoft tool section to the coding guidelines page (#36911)
zephyrproject-rtos/zephyr: doc/contribute/coding_guidelines/index.rst

      */
     /* parasoft suppress item Rule ID */

  Below you can find an example of a recommended commit message::

     testsuite: suppress usage of setjmp in a testcode (rule 21.4)

     According to the Rule 21.4 the standard header file <setjmp.h> shall not
     be used. Suppress it, because it raises violation in a testcode,
     not in a runtime code. Tag suppresses reporting of violation for the
     current file, starting from the line where the suppression is located.

     Found as a coding guideline violation (Rule 21.4) by static coding
     scanning tool.

     Signed-off-by: Johnny Developer johnny.developer@company.com

  The example below demonstrates how false positives can be suppressed::

     /* Static code analysis tool can raise a violation that the standard header
      * <setjmp.h> shall not be used.
      *
      * setjmp is using in a test code, not in a runtime code
      */
     /* parasoft suppress item MISRAC2012-RULE_21_4-a */
     /* parasoft suppress item MISRAC2012-RULE_21_4-b */

  The Parasoft tags will suppress all violations of the specific rule until
  the end of the file. To read more about Suppression Findings in the

>   Parasoft tool refer to the official Parasoft `documentation`_

Reviewer: >suppress all violations of the specific rule until
  the end of the file

Is this really what we want?

As per the [documentation](https://docs.parasoft.com/display/CPPTEST1031/Suppressing+Findings), `parasoft suppress` and `parasoft-suppress` are not the same -- the former does what is described here, whereas the latter only suppresses the specified rule for the current (following) line.

FWIW, I think the scope of suppression should be limited to the place where the `parasoft` tag is specified; otherwise, this might have an unintended consequence of suppressing other unsanctioned violations in the same file.

cc @AmberHibberd @ioannisg

Reviewer: Also, how does Parasoft Codescan handle multiple suppressions with `parasoft-suppress`?

For instance, if we have:
```
/* parasoft-suppress item MISRAC2012-RULE_21_4-a */
/* parasoft-suppress item MISRAC2012-RULE_21_4-b */
SOME CODE
```

Are both `MISRAC2012-RULE_21_4-a` and `MISRAC2012-RULE_21_4-b` suppressed for `SOME CODE`? or only `MISRAC2012-RULE_21_4-b`?

Author: Yes, about suppression range necessary to think together. About comment, I think better to have standalone comment.

Reviewer: > I think better to have standalone explanation comment.

Sure.
```
/* parasoft-suppress MISRAC2012-RULE_21_4-a "Comment explaining rationale behind this suppression" */
/* parasoft-suppress MISRAC2012-RULE_21_4-b "Comment explaining rationale behind this suppression" */
SOME CODE
```

Author: I don't know if it supports doing like that. @ceolin told me, he has access to the Parasoft directly and can try to share with me an access. I can only access dashboard with the results, but can't run tool to see how it detects such kind of tags.

Author: > > I think better to have standalone explanation comment.
> 
> Sure (assuming Parasoft Codescan supports the following).
> 
> ```
> /* parasoft-suppress MISRAC2012-RULE_21_4-a "Comment explaining rationale behind this suppression" */
> /* parasoft-suppress MISRAC2012-RULE_21_4-b "Comment explaining rationale behind this suppression" */
> SOME CODE
> ```

I don't know if it supports doing like that. @ceolin told me, he has access to the Parasoft directly and can try to share with me an access. I can only access dashboard with the results, but can't run tool to see how it detects such kind of tags.

Author: @stephanosio you can search on Github "parasoft-suppress", it can give a baseline for use tags, like other developers are doing. I'm reviewing it right now.

Author: @stephanosio Updated my PR.
1. You are right, better to use parasoft tag that can suppress only 1 line of the code.
`violation line of the code /* parasoft-suppress MISRAC2012-RULE_21_4-a*/`

I see that other developers keep that tag in one line even if it is too long.

2. According to the idea adding a tag with a comment:
`CODE LINE parasoft-suppress MISRAC2012-RULE_21_4-b "Comment explaining rationale behind this suppression" */`

Explanation comment can be really long, and probably it can't fit into one line. I see that no one is doing multiline comments using parasoft tags. So I suggest better keep comment separate (above the violation code line) and the comment should have the same style as other comments in the Zephyr project.

Author: Also I created a forum topic on the Parasoft webpage. But seems their forum is dead...
https://forums.parasoft.com/discussion/5320/multiline-parasoft-tags-to-suppress-violations

Reviewer 2: I agree keeping the rationale separated of the suppression.

  .. _documentation: https://docs.parasoft.com/display/CPPTEST1031/Suppressing+Findings

Step 5

>   After a PR is submitted a developer needs to add the ``Coding guidelines`` and

Reviewer: ```suggestion
  After a PR is submitted, a developer needs to add the ``Coding guidelines`` and
```

>   ``MISRA-C`` Github labels, so his PR can be easily tracked by maintainers.

Reviewer: Nit: political correctness
```suggestion
  ``MISRA-C`` Github labels, so their PR can be easily tracked by maintainers.
```

and precision in bug findings.

Violations fixing process
=========================

Step 1
  Any Zephyr Project member, company or a developer can request a license
  to access Codescan online dashboard.

Step 2
  A developer starts to review raised violations.

Step 3
  A developer submits a Github PR with the fix. Commit messages should follow
  the same guidelines as other PRs in the Zephyr project. Please add a comment
  that your fix was found by a static coding scanning tool.
  A developer should follow and refer to the Zephyr :ref:`coding_guidelines`
  as the basic coding rules. These rules are based on the MISRA C standard.

  Below you can find an example of a recommended commit message::

     lib: os: add braces to 'if' statements

     An 'if' (expression) construct shall be followed by a compound statement.
     Add braces to improve readability and maintainability.

     Found as a coding guideline violation (Rule 15.6) by static
     coding scanning tool.

     Signed-off-by: Johnny Developer johnny.developer@company.com

Step 4
  If a violation is a false-positive developer must submit a PR with
  a suppression tag and an explanation comment.

> 

Reviewer 2: IMHO the suppression should not used for false-positives but for deviations. Deviations are real violations that *we* deliberated decided to ignore for a reason (that should be explained in the comment). False positives should just be marked in the tool just like we do for coverity.

I don't like the idea of adding comments in the code due a tool problem.

Author: We will suppress false positives and deviations. All code false positive violations should be marked inside of the code to be ignored by the tool. All deviations should be marked in the code too with the explanation comment. We shouldn't rely on the tool, so standalone explanation comment can help. Even if we will remove tags, comments will stay in the code.

      * a false-positive. Don't refer to the Parasoft tool here, just mention
      * that static code analysis tool can raise/raised a violation
      * in the line below.
      */
     code_line_with_a_violation /* parasoft-suppress Rule ID */

  Below you can find an example of a recommended commit message::

     testsuite: suppress usage of setjmp in a testcode (rule 21.4)

     According to the Rule 21.4 the standard header file <setjmp.h> shall not
     be used. Suppress it, because it raises violation in a testcode,
     not in a runtime code. Tag suppresses reporting of violation for the
     current file, starting from the line where the suppression is located.

     Found as a coding guideline violation (Rule 21.4) by static coding
     scanning tool.

     Signed-off-by: Johnny Developer johnny.developer@company.com

  The example below demonstrates how false positives can be suppressed in the code::

     /* Static code analysis tool can raise a violation that the standard header
      * <setjmp.h> shall not be used.
      *
      * setjmp is using in a test code, not in a runtime code
      */

>      #include <setjmp.h> /* parasoft-suppress MISRAC2012-RULE_21_4-a MISRAC2012-RULE_21_4-b */

Reviewer: The line length for this example already exceeds 80.

Should these be exceptions to the "80 column rule"? If not, how shall we handle line changes? Can Parasoft Codescan handle line change midst a `parasoft-suppress` statement?

Author: > The line length for this example already exceeds 80.
> 
> Should these be exceptions to the "80 column rule"? If not, how shall we handle line changes? Can Parasoft Codescan handle line change midst a `parasoft-suppress` statement?

I'm trying to receive access to the Parasoft tool. I don't have online or desktop versions of the Parasoft to test it.
I think, if we will use `parasoft-suppress` tag, necessary to make an exception. Because that tag was designed to suppress a line were the tag itself was placed. We will often face with the situations where line with a tag will exceed 80 symbols.
Currently I don't have PRs that will have such situations if I will apply a new kind of a tag
https://github.com/zephyrproject-rtos/zephyr/pull/36772/files
https://github.com/zephyrproject-rtos/zephyr/pull/36771/files
https://github.com/zephyrproject-rtos/zephyr/pull/36773/files

It will be a problem, only if there will be a single line explanation comment. But as you can see, all comments are placed above line with a violation.

.. _OSHWA Resolution to Redefine SPI Signal Names: https://www.oshwa.org/a-resolution-to-redefine-spi-signal-names/
.. _Zephyr issue 27033: https://github.com/zephyrproject-rtos/zephyr/issues/27033

Parasoft Codescan Tool
**********************

Parasoft Codescan is an official static code analysis tool of the Zephyr
project. It is used to automate compliance with a range of coding and security
standards.
Currently, the tool is set to MISRA C:2012 Coding Standard because Zephyr
`Coding Guidelines`_ are based on that standard.
It is used together with the Coverity Scan tool to achieve the best code health
and precision in bug findings.

Violations fixing process
=========================
Step 1
  Zephyr Project member company developer can request a license to access
  Codescan online dashboard.

Step 2
  A developer is starting to review raised violations.

Step 3
  A developer submits a Github PR with the fix. Commit messages should follow
  the same guidelines as other PRs in the Zephyr project. Please add a comment
  that your fix was found by a static coding scanning tool.
  A developer should follow and refer to the Zephyr `Coding Guidelines`_ as the

>   basic coding rules. That rules are based on the MISRA C standard.

Reviewer 3: ```suggestion
  basic coding rules. These rules are based on the MISRA C standard.
```

>   .. _Coding Guidelines: https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html
> 
> 

Reviewer 3: ```suggestion
```

  We recommend following that example PR::

     lib: os: add braces to 'if' statements

     An 'if' (expression) construct shall be followed by a compound statement.
     Add braces to improve readability and maintainability.

     Found as a coding guideline violation (Rule 15.6) by static
     coding scanning tool.

     Signed-off-by: Johnny Developer johnny.developer@company.com

Step 4
  If a violation is a false-positive developer must submit a PR with

>   a suppression tag and explanation comment.

Reviewer 3: ```suggestion
  a suppression tag and an explanation comment.
```

  We recommend following that example PR::

     testsuite: suppress usage of setjmp in a testcode (rule 21.4)

     According to the Rule 21.4 the standard header file <setjmp.h> shall not
     be used. Suppress it, because it raises violation in a testcode,
     not in a runtime code. Tag suppresses reporting of violation for the
     current file, starting from the line where the suppression is located.

     Found as a coding guideline violation (Rule 21.4) by static coding
     scanning tool.

     Signed-off-by: Johnny Developer johnny.developer@company.com

>   We recommend following that example to add a comment and tag to suppress
>   a violation::

Reviewer 3: ```suggestion
  The example below demonstrates how false positives can be suppressed::
```

     /* Static code analysis tool can raise a violation that the standard header
      * <setjmp.h> shall not be used.
      *
      * setjmp is using in a test code, not in a runtime code
      */
     /* parasoft suppress item MISRAC2012-RULE_21_4-a */
     /* parasoft suppress item MISRAC2012-RULE_21_4-b */

  The template structure of the comment and tag in the code should be::

     /* Explain why that part of the code is healthy and raised violation is
      * a false-positive
      */
     /* parasoft suppress item Rule ID */

  Added Parasoft tag will suppress all violations of the specific rule till
  the end of the file. To read more about Suppression Findings in the
  Parasoft tool refer to the official Parasoft `documentation`_

>   .. _documentation: http://www.sphinx-doc.org/en/stable/contents.html

Reviewer 3: This link is not correct.

Step 5

>   After a PR were submitted a developer needs to add ``Coding guidelines`` and

Reviewer 3: ```suggestion
  After a PR is submitted a developer needs to add the ``Coding guidelines`` and
```

  ``MISRA-C`` Github labels, so his PR can be easily tracked by maintainers.
  If you have any concerns about what your PR should look like, you can search

>   on Github by the tags and find similar PRs that already got merged.

Reviewer 3: ```suggestion
  on Github using the mentioned tags and find similar PRs that already got merged.
```

Parasoft Codescan Tool
**********************

Parasoft Codescan is an official static code analysis tool used by the Zephyr
project. It is used to automate compliance with a range of coding and security
standards.
Currently, the tool is set to MISRA C:2012 Coding Standard because Zephyr
:ref:`coding_guidelines` are based on that standard.
It is used together with the Coverity Scan tool to achieve the best code health
and precision in bug findings.

Violations fixing process
=========================

Step 1
  Any Zephyr Project member, company or a developer can request a license
  to access Codescan online dashboard.

Step 2
  A developer starts to review raised violations.

Step 3
  A developer submits a Github PR with the fix. Commit messages should follow
  the same guidelines as other PRs in the Zephyr project. Please add a comment
  that your fix was found by a static coding scanning tool.
  A developer should follow and refer to the Zephyr :ref:`coding_guidelines`
  as the basic coding rules. These rules are based on the MISRA C standard.

  Below you can find an example of a recommended commit message::

     lib: os: add braces to 'if' statements

> 

Reviewer 4: ```suggestion
```

     An 'if' (expression) construct shall be followed by a compound statement.
     Add braces to improve readability and maintainability.

     Found as a coding guideline violation (Rule 15.6) by static
     coding scanning tool.

>      Signed-off-by: Johnny Developer johnny.developer@company.com

Reviewer 4: ```suggestion
     Signed-off-by: Johnny Developer <johnny.developer@company.com>
```

Step 4
  If a violation is a false-positive developer should mark it in the Codescan
  tool just like we do for the Coverity tool.
  Also developer should add an explanation comment in the code, explaining that
  static code analysis tool may raise a violation there, but consider it as
  a false-positive.

Step 5
  If a developer found a real violation that community decided to ignore
  for a reason, developer must submit a PR with a suppression tag
  and an explanation comment.
  The template structure of the comment and tag in the code should be::

     /* Explain why that part of the code doesn't follow the standard,
      * explain why it is a deliberated deviation from the standard.
      * Don't refer to the Parasoft tool here, just mention that static code
      * analysis tool raised a violation in the line below.
      */
     code_line_with_a_violation /* parasoft-suppress Rule ID */

  Below you can find an example of a recommended commit message::

     testsuite: suppress usage of setjmp in a testcode (rule 21.4)

> 

Reviewer 4: ```suggestion
```

     According to the Rule 21.4 the standard header file <setjmp.h> shall not
     be used. Suppress it, because it raises violation in a testcode,
     not in a runtime code. Tag suppresses reporting of the violation for the
     current the line where the suppression is located.
     It is a deviberated deviation.

     Found as a coding guideline violation (Rule 21.4) by static coding
     scanning tool.

>      Signed-off-by: Johnny Developer johnny.developer@company.com

Reviewer 4: ```suggestion
     Signed-off-by: Johnny Developer <johnny.developer@company.com>
```
